File: 2023/3/main.py
import re
import time

i = open(0).read()
w = 1 + i.find("\n")

# Padding the input removes edge cases!
pad = "." * (w - 1) + "\n"
i = 2*pad + i + 2*pad
print(i)


integer = re.compile(r"\d+")
part = re.compile(r"[^\d\.\n]")
gear = re.compile(r"\*")


# Part One
s, numbers = 0, integer.finditer(i)
for m in numbers:
    # Where is the number?
    start = m.start() - 1
    end = m.end() + 1
    
    # The "box" surrounding the number.
    box = i[start-w:end-w] + i[start:end] + i[start+w:end+w]
    if part.search(box):
        s += int(m[0])

print(s)
    
    
# Part Two
t0 = time.perf_counter_ns()

ratios, gears = 0, gear.finditer(i)
for m in gears:
    # Where is the gear?
    pos = m.start()

    # Get the numbers in ±1 row.
    start = pos - (pos % w) - w
    end = start + 3 * w
    nums = integer.finditer(i, pos=start, endpos=end)

    p = 1
    j = 0
    for n in nums:
        s = max(n.start() % w - 1, 0)
        e = n.end() % w 
        pos = pos % w
        if s <= pos <= e:
            p *= int(n[0])
            j += 1
            
    if j == 2:
        ratios += p

t1 = time.perf_counter_ns()
delta = t1 - t0
        
print(ratios)
print(f"Δt={delta}ns")

File: 2023/3/main.ml
open Aoc
open Aoc.IO

let symbol = Re.compiled "[^\\d\\.\\n]"
let integer = Re.compiled "\\d+"
let gear = Re.compiled "\\*"
    
let sum xs = List.fold_right ( + ) xs 0
let product xs = List.fold_right ( * ) xs 1
    
let p1 input width =
  let numbers = Re.all integer input in
  
  let part_nos =
    numbers |>
    List.map
      (fun g ->
         (* Where is the number? *)
         let start = Re.start g in
         let stop = Re.stop g in

         (* "Columns" we are searching between *)
         let start = max (start - 1) 0 in
         let len = stop - start + 1 in

         (* Are there any symbols here? *)
         let box = String.sub input  start          len ^
                   String.sub input (start - width) len ^
                   String.sub input (start + width) len in
         let is_part = Re.execp symbol box in

         let part_no = Re.int (Re.str g) in
         
         is_part, part_no) |>
    List.filter fst |>
    List.map snd in

  Fmt.(pr "Part One: @{<green>%i@}\n") (sum part_nos)

                

let p2 input width =
  let gears = Re.all gear input in
  
  let surrounding_nos =
    gears |> List.map
      (fun g ->
         (* Where is the gear? *)
         let pos = Re.start g in

         (* The numbers in ±1 row. *)
         let start = pos - (pos mod width) - width in
         let len = 3 * width in
         let nums = Re.all ~pos:start ~len:len integer input in
         
         (* Discard numbers not next to the gear. *)
         nums |>
         List.filter
           (fun g ->
              let s = max (Re.start g mod width - 1) 0 in
              let e = Re.stop g mod width in
              let pos = pos mod width in 
              s <= pos && pos <= e) |>
         (* Extract the number from the regex group. *)
         List.map Re.str |>
         List.map Re.int) |>
    (* Keep gears with exactly 2 surrounding numbers. *)
    List.filter (fun adj -> List.length adj = 2) |>
    (* Calculate the gear ratio. *)
    List.map (fun adj -> product adj) in
  
  Fmt.(pr "Part Two: @{<green>%i@}\n") (sum surrounding_nos)

let () = 
  let input = read stdin in
  let width = 1 + Str.find ~sub:"\n" input in

  (* Padding the input removes some edge cases. *)
  let pad   = (Str.repeat "." (width - 1)) ^ "\n" in
  let input = pad ^ pad ^ input ^ pad ^ pad in
  
  p2 input width;
